tfhe/src/integer/server_key/radix_parallel/ilog2.rs

```diff
@@ -0,0 +1,734 @@
+        // or becomes 0 if the preceding block
+        // had a bit set to one in it (leading_zeros != num bits in message)
+        let num_bits_in_message = self.key.message_modulus.0.ilog2() as u64;
+        let sum_lut = self.key.generate_lookup_table_bivariate(
+            |block_num_leading_zeros, more_significant_block_leading_zeros| {
+                if more_significant_block_leading_zeros == num_bits_in_message {
+                    block_num_leading_zeros
+                } else {
+                    0
+                }
+            },
+        );
+
+        let sum_function =
+            |block_num_leading_zeros: &mut Ciphertext,
+             more_significant_block_leading_zeros: &Ciphertext| {
+                self.key.unchecked_apply_lookup_table_bivariate_assign(
+                    block_num_leading_zeros,
+                    more_significant_block_leading_zeros,
+                    &sum_lut,
+                );
+            };
+        self.compute_prefix_sum_hillis_steele(blocks, sum_function)
+    }
+
+    /// Counts how many consecutive bits there are
+    fn count_consecutive_bits<T>(
+        &self,
+        ct: &T,
+        direction: Direction,
+        bit_value: BitValue,
+    ) -> RadixCiphertext
+    where
+        T: IntegerRadixCiphertext,
+    {
+        if ct.blocks().is_empty() {
+            return self.create_trivial_zero_radix(ct.blocks().len());
+        }
+
+        let num_bits_in_message = self.key.message_modulus.0.ilog2();
+        let original_num_blocks = ct.blocks().len();
+
+        let num_bits_in_ciphertext = num_bits_in_message
+            .checked_mul(original_num_blocks as u32)
+            .expect("Number of bits encrypted exceeds u32::MAX");
+
+        let leading_count_per_blocks =
+            self.prepare_count_of_consecutive_bits(ct.clone(), direction, bit_value);
+
+        // `num_bits_in_ciphertext` is the max value we want to represent
+        // its ilog2 + 1 gives use how many bits we need to be able to represent it.
+        let counter_num_blocks = divide_ceil(
+            num_bits_in_ciphertext.ilog2() + 1 + u32::from(T::IS_SIGNED),
+            self.message_modulus().0.ilog2(),
+        );
+
+        let cts = leading_count_per_blocks
+            .into_iter()
+            .map(|block| {
+                let mut ct: RadixCiphertext =
+                    self.create_trivial_zero_radix(counter_num_blocks as usize);
+                ct.blocks[0] = block;
+                ct
+            })
+            .collect::<Vec<_>>();
+
+        self.unchecked_sum_ciphertexts_vec_parallelized(cts)
+            .expect("internal error, empty ciphertext count")
+    }
+
+    //==============================================================================================
+    //  Unchecked
+    //==============================================================================================
+
+    pub fn unchecked_trailing_zeros<T>(&self, ct: &T) -> RadixCiphertext
```

Comment:
let's add a "check doc of default_function" with a link, it's not perfect but at least the function re-uses documentation from elsewhere,

same for all functions missing a docstring, I think we have some ready to use wording for unchecked and smarts ops somewhere (maybe bitops or comparisons have that)